# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/device/gl.rs
# Commit: 0415da2e9198
# Full Hash: 0415da2e9198c74c8881269e600dca28f38b8819
# Author: Glenn Watson <gw@intuitionlibrary.com>
# Date: 2020-04-14 04:09:24
# Regressor Bug: 1629693
# File Overlap Count: 1
# Description:
#   Bug 1629693 - Enable GPU cache scatter path when running ANGLE. r=kvark
#   
#   On some (mostly older, integrated) GPUs, the normal GPU texture cache
#   update path doesn't work well when running on ANGLE, causing CPU stalls
#   inside D3D and/or the GPU driver. To reduce the number of code paths we
# ==============================================================================

diff -r 17c0b08918ac -r 0415da2e9198 gfx/wr/webrender/src/device/gl.rs
--- a/gfx/wr/webrender/src/device/gl.rs	Mon Apr 13 21:38:51 2020 +0000
+++ b/gfx/wr/webrender/src/device/gl.rs	Mon Apr 13 23:26:07 2020 +0000
@@ -790,7 +790,7 @@
         let source_and_digest = SHADERS.get(&name).expect("Shader not found");
 
         // Hash the renderer name.
-        hasher.write(device.renderer_name.as_bytes());
+        hasher.write(device.capabilities.renderer_name.as_bytes());
 
         // Hash the prefix string.
         build_shader_prefix_string(
@@ -999,6 +999,8 @@
     pub supports_nonzero_pbo_offsets: bool,
     /// Whether the driver supports specifying the texture usage up front.
     pub supports_texture_usage: bool,
+    /// The name of the renderer, as reported by GL
+    pub renderer_name: String,
 }
 
 #[derive(Clone, Debug)]
@@ -1083,7 +1085,6 @@
 
     max_texture_size: i32,
     max_texture_layers: u32,
-    renderer_name: String,
     cached_programs: Option<Rc<ProgramCache>>,
 
     // Frame counter. This is used to map between CPU
@@ -1554,6 +1555,7 @@
                 supports_texture_swizzle,
                 supports_nonzero_pbo_offsets,
                 supports_texture_usage,
+                renderer_name,
             },
 
             color_formats,
@@ -1578,7 +1580,6 @@
 
             max_texture_size,
             max_texture_layers,
-            renderer_name,
             cached_programs,
             frame_id: GpuFrameId(0),
             extensions,
@@ -1998,7 +1999,7 @@
                     error!(
                       "Failed to load a program object with a program binary: {} renderer {}\n{}",
                       &info.base_filename,
-                      self.renderer_name,
+                      self.capabilities.renderer_name,
                       error_log
                     );
                     if let Some(ref program_cache_handler) = cached_programs.program_cache_handler {